### PR TITLE
Make Cancel-able HTTP Requests

### LINKED ISSUE #109: Make Cancel-able HTTP Requests
Currently, when you press the Send Button, a request is sent and the UI waits for it to complete.
In case no response is sent for a very long time (server might hang), there should be a provision to cancel the request instead of restarting the entire app.

### PR DESCRIPTION
**Pull request by yashas-hm**
### Problem
Natively the http package doesn't support cancellation of requests. 

### Solution 
I installed the Dio package which provides an option to cancel ongoing requests. I have migrated all the http requests to Dio and also modified the response model accordingly.

### UI Changes
I have added a cancel option in place of the sending button and also added a cancelled screen similar to not sent when user cancels a request. I have modified the button to provide visual changes:
1. Button turns red for cancel request.
2. Button turns blue for send request.
3. Button turns grey for busy status. 

## Related Issues
- Related Issue #109  
- Closes #109 

### Checklist
- [x] I have gone through the [contributing guide](https://github.com/foss42/apidash/blob/main/CONTRIBUTING.md)
- [x] I have updated my branch and synced it with project `main` branch before making this PR
- [x] I have run the tests (`flutter test`) and all tests are passing

## Added/updated tests?
- [x] Yes
- [ ] No, and this is why:

### Preview
Screenshot 2024-02-15 at 7 19 04 PM
Screenshot 2024-02-15 at 7 19 12 PM
Screenshot 2024-02-15 at 7 19 36 PM
Screenshot 2024-02-15 at 7 19 52 PM

PS:
This is my second open source contribution. Please do tell me what can be improved. Documentation wise, commits wise and code wise.

Thanks

### CODE REVIEW

**Comment by ashitaprasad** [comment]
Thanks for sending the PR @yashas-hm 
I'll review it and post my feedback soon.

**Comment by yashas-hm** [comment]
Hi any update?

Thanks

**Comment by animator** [comment]
@yashas-hm This PR is under review. We will post updates after completing the review.

**Comment by ashitaprasad** [comment]
@yashas-hm We reviewed the PR. It is neither cancelling the request nor we are obtaining any response. Can you take a look at it again.

**Comment by yashas-hm** [comment]
@ashitaprasad the updates from main that I had fetched after raising the PR caused the functionality to break will fix and push changes by tomorrow.

**Comment by ashitaprasad** [comment]
cool @yashas-hm.

**Comment by yashas-hm** [comment]
Fixed all issues please check now @ashitaprasad

**Comment by ashitaprasad** [comment]
Some of the tests are failing. Kindly relook.

File: lib/models/request_model.dart [hunk 1]
```diff
@@ -88,6 +91,7 @@ class RequestModel {
   bool get hasFileInFormData => formDataList
       .map((e) => e.type == FormDataType.file)
       .any((element) => element);
+  bool get isFormDataRequest => requestBodyContentType == ContentType.formdata;
```
[the remark this answers is not in the document]
Comment (yashas-hm) [reply lib/models/request_model.dart:94]: Ok will do and push changes.

**Comment by yashas-hm** [comment]
@ashitaprasad  Fixed the issues  👍🏻

**Comment by yashas-hm** [comment]
@ashitaprasad @animator please review

**Comment by yashas-hm** [comment]
Yes I did but its is based on an older version of http package(1.0.0) whereas the current version is 1.2.1. Also Dio give us more control over the data and how requests are sent. Additionally Dio is more maintained than cancellation_token_http.

Dio provides a lot of extra functionality compared to Http. For example, Dio supports interceptors, which allow you to perform actions before and after each request. This can be useful for logging, error handling, and other tasks if it is needed to be implemented in future.

**Comment by animator** [comment]
We are currently not planning to migrate to Dio. If we do so in future will re-open this issue.

**Pull request closed without merging**